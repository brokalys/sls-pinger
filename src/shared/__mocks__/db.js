export const query = jest.fn();
export const getPingersByFrequency = jest.fn().mockReturnValue([]);
export const getPropertyQueueForPingers = jest.fn().mockReturnValue([]);
export const lockPropertyQueueItems = jest.fn();
export const deletePropertyQueueItems = jest.fn();
export const getEmailsThatShouldBeLimitLocked = jest.fn().mockReturnValue([]);
export const getEmailsWithLimitLockerNotification = jest
  .fn()
  .mockReturnValue([]);
export const limitLockPingerEmails = jest.fn();
export const createPingerStatsEntry = jest.fn();
export const logPingerAttempt = jest.fn();
export const updatePingerAttemptTimestamp = jest.fn();
export const getAvailablePingers = jest.fn().mockReturnValue([]);
export const queuePingerForSummaryEmail = jest.fn();
export const getPropertyStats = jest.fn().mockReturnValue([]);
