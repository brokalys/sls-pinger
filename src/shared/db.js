import serverlessMysql from 'serverless-mysql';
import moment from 'moment';

const connection = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    database: process.env.DB_DATABASE,
    timezone: 'Z',
    typeCast: true,
  },
});

export function getPingersByType(type) {
  return connection.query({
    sql: `
      SELECT *
      FROM pinger_emails
      WHERE unsubscribed_at IS NULL
        AND (limit_reached_at IS NULL OR limit_reached_at < ? OR is_premium = true)
        AND type = ?
   `,
    values: [moment.utc().startOf('month').toDate(), type],
    typeCast(field, next) {
      if (field.type === 'TINY' && field.length === 1) {
        return field.string() === '1';
      }

      if (field.name === 'categories' || field.name === 'types') {
        return JSON.parse(field.string());
      }

      return next();
    },
  });
}

export function getPropertyQueueForPingers(pingerIds) {
  return connection.query({
    sql: `
      SELECT *
      FROM pinger_queue
      WHERE pinger_id IN (?)
        AND locked = 0
   `,
    values: [pingerIds],
  });
}

export function lockPropertyQueueItems(itemIds) {
  return connection.query({
    sql: `
      UPDATE pinger_queue
      SET locked = 1
      WHERE id IN (?)
   `,
    values: [itemIds],
  });
}

export function deletePropertyQueueItems(itemIds) {
  return connection.query({
    sql: `
      DELETE FROM pinger_queue
      WHERE id IN (?)
   `,
    values: [itemIds],
  });
}
