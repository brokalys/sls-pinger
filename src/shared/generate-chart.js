const vega = require('vega');
const chartBase = require('./data/base-chart.json');
const s3 = require('./s3');

module.exports = async function generateChart(fileName, values, maxDate) {
  chartBase.data[0].values = values;
  chartBase.scales[0].domainMax = maxDate;

  const view = new vega.View(vega.parse(chartBase), {
    renderer: 'none',
  });

  const svgStr = await view.toSVG();
  const bucketName = `${process.env.STAGE}-pinger-charts`;

  await s3
    .putObject({
      Bucket: bucketName,
      Key: fileName,
      Body: svgStr,
      ContentType: 'image/svg+xml',
    })
    .promise();

  return `https://${bucketName}.s3.${process.env.AWS_REGION}.amazonaws.com/${fileName}`;
};
