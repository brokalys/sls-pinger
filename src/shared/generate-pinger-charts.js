const moment = require('moment');
const db = require('./db');
const generateChart = require('./generate-chart');

module.exports = async function generatePingerCharts(pingers) {
  const pingerIds = pingers.map(({ id }) => id);
  const pingerHashMap = pingers.reduce(
    (carry, pinger) => ({ ...carry, [pinger.id]: pinger }),
    {},
  );
  const stats = await db.getPropertyStats(pingerIds);

  const state = {
    ids: [],
    entities: {},
  };

  stats.forEach((row) => {
    const createdAt = moment(row.created_at).format('YYYY-MM-DD');

    if (!state.entities[row.pinger_id]) {
      state.entities[row.pinger_id] = {
        values: [],
        minDate: null,
        maxDate: null,
        uniqueDates: [],
        frequency: pingerHashMap[row.pinger_id].frequency,
      };
      state.ids.push(row.pinger_id);
    }

    if (!state.entities[row.pinger_id].uniqueDates.includes(createdAt)) {
      state.entities[row.pinger_id].uniqueDates.push(createdAt);
    }

    if (
      !state.entities[row.pinger_id].minDate ||
      state.entities[row.pinger_id].minDate > row.created_at
    ) {
      state.entities[row.pinger_id].minDate = row.created_at;
    }

    if (
      !state.entities[row.pinger_id].maxDate ||
      state.entities[row.pinger_id].maxDate < row.created_at
    ) {
      state.entities[row.pinger_id].maxDate = row.created_at;
    }

    row.data.forEach((price) => {
      state.entities[row.pinger_id].values.push({
        x: createdAt,
        y: price,
      });
    });
  });

  const idsWithCharts = state.ids.filter(
    (id) => state.entities[id].uniqueDates.length >= 2,
  );

  const urls = await Promise.all(
    idsWithCharts.map((id) => {
      const idHash = pingerHashMap[id].id_hash;
      const maxDate = moment(state.entities[id].maxDate);
      const calculatedFutureDate = moment(state.entities[id].minDate).add(
        11,
        mapPingerFrequencyToMomentJs(state.entities[id].frequency),
      );

      return generateChart(
        `${idHash}.svg`,
        state.entities[id].values,
        (maxDate.isAfter(calculatedFutureDate)
          ? maxDate
          : calculatedFutureDate
        ).valueOf(),
      );
    }),
  );

  return urls.reduce(
    (carry, url, index) => ({ ...carry, [idsWithCharts[index]]: url }),
    {},
  );
};

function mapPingerFrequencyToMomentJs(frequency) {
  switch (frequency) {
    case 'daily':
      return 'days';
    case 'weekly':
      return 'weeks';
    case 'monthly':
      return 'months';
  }

  return 'days';
}
