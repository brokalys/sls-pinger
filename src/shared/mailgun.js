import mailgun from 'mailgun-js';

export default mailgun({
  apiKey: process.env.MAILGUN_API_KEY,
  domain: process.env.MAILGUN_DOMAIN,
});
