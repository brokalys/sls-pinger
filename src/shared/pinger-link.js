function createUnsubscribeLink(pinger) {
  return `https://unsubscribe.brokalys.com/?key=${encodeURIComponent(
    pinger.unsubscribe_key,
  )}&id=${encodeURIComponent(pinger.id_hash)}`;
}

function createAllPingersLink(pinger) {
  const id = encodeURIComponent(pinger.id_hash);
  const unsubscribeKey = encodeURIComponent(pinger.unsubscribe_key);

  return `https://pinger.brokalys.com/#/pingers/${id},${unsubscribeKey}`;
}

module.exports = {
  unsubscribeLink: createUnsubscribeLink,
  allPingersLink: createAllPingersLink,
};
