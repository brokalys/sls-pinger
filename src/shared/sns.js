const AWS = require('aws-sdk');

AWS.config.update({ region: process.env.AWS_REGION });

module.exports = new AWS.SNS({ apiVersion: '2010-03-31' });
