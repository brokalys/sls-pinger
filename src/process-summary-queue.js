import * as db from './shared/db';
import generatePingerCharts from './shared/generate-pinger-charts';
import sns from './shared/sns';
import createUnsubscribeLink from './shared/unsubscribe-link';

export async function run(event, context = {}) {
  const { frequency = 'daily' } = event;
  context.callbackWaitsForEmptyEventLoop = false;

  // Retrieve all the PINGERS with frequency {variable} from the DB
  const pingers = await db.getPingersByFrequency(frequency);
  const pingerIds = pingers.map(({ id }) => id);

  if (pingerIds.length <= 0) {
    return;
  }

  // Retrieve the queue for all the matched pingers
  const propertyQueue = await db.getPropertyQueueForPingers(pingerIds);
  const propertyQueueIds = propertyQueue.map(({ id }) => id);

  // Create a map for properties in each PINGER
  // Sample: { 123: [{ ... }, { ... }] }
  const properties = propertyQueue.reduce(
    (carry, data) => ({
      ...carry,
      [data.pinger_id]: [
        ...(carry[data.pinger_id] || []),
        JSON.parse(data.data),
      ],
    }),
    {},
  );

  // Filter PINGERS that have properties
  const pingersWithProperties = pingers.filter((data) => !!properties[data.id]);

  // What if no matches?
  if (pingersWithProperties.length === 0) {
    return;
  }

  // Mark property queue entries as "locked" so we don't
  // send duplicate emails in case of an error
  await db.lockPropertyQueueItems(propertyQueueIds);

  // Insert a stats entry for each pinger
  await Promise.all(
    pingersWithProperties.map((pinger) =>
      db.createPingerStatsEntry(
        pinger.id,
        properties[pinger.id].map(
          ({ calc_price_per_sqm }) => calc_price_per_sqm,
        ),
      ),
    ),
  );

  // Generate summary images for each pinger
  const urls = await generatePingerCharts(pingerIds);

  // Publish SNS notification to send email for each PINGER that has properties
  await Promise.all(
    pingersWithProperties.map((pinger) =>
      sendEmail(pinger, properties[pinger.id], urls[pinger.id]),
    ),
  );

  // Delete the property queue from DB
  await db.deletePropertyQueueItems(propertyQueueIds);
}

function sendEmail(pinger, properties, heroImgUrl) {
  return sns
    .publish({
      Message: 'email',
      MessageAttributes: {
        to: {
          DataType: 'String',
          StringValue: pinger.email,
        },
        subject: {
          DataType: 'String',
          StringValue: 'Jauni PINGER sludinājumi',
        },
        pinger_id: {
          DataType: 'Number',
          StringValue: String(pinger.id),
        },
        template_id: {
          DataType: 'String',
          StringValue: 'summary',
        },
        template_variables: {
          DataType: 'String',
          StringValue: JSON.stringify({
            is_premium: pinger.is_premium,
            hero_img_url: heroImgUrl,
            unsubscribe_url: createUnsubscribeLink(pinger),
            properties: properties.map((data) => [
              data.url,
              data.price,
              data.rooms,
              data.area,
            ]),
          }),
        },
      },
      MessageStructure: 'string',
      TargetArn: `arn:aws:sns:${process.env.AWS_REGION}:173751334418:email-${process.env.STAGE}`,
    })
    .promise();
}
