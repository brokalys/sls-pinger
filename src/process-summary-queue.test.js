import * as db from './shared/db';
import sns from './shared/sns';
import { run } from './process-summary-queue';
import {
  createPingerFixture,
  createPropertyQueueItemFixture,
} from '__fixtures__';

jest.mock('./shared/db');
jest.mock('./shared/sns');

describe('process-summary-queue', () => {
  afterEach(jest.clearAllMocks);

  test('sends an email SNS notification', async () => {
    db.getPingersByType.mockReturnValue([createPingerFixture()]);
    db.getPropertyQueueForPingers.mockReturnValue([
      createPropertyQueueItemFixture({ id: 1 }),
      createPropertyQueueItemFixture({ id: 2 }),
      createPropertyQueueItemFixture({ id: 3 }),
    ]);

    await run({ type: 'daily' });

    expect(sns.publish).toBeCalledTimes(1);
  });

  test('does not send a SNS notification if there are no PINGER', async () => {
    db.getPingersByType.mockReturnValue([]);
    db.getPropertyQueueForPingers.mockReturnValue([
      createPropertyQueueItemFixture({ id: 1 }),
      createPropertyQueueItemFixture({ id: 2 }),
      createPropertyQueueItemFixture({ id: 3 }),
    ]);

    await run({ type: 'daily' });

    expect(sns.publish).toBeCalledTimes(0);
  });

  test('does not send a SNS notification if there are no properties', async () => {
    db.getPingersByType.mockReturnValue([createPingerFixture()]);
    db.getPropertyQueueForPingers.mockReturnValue([]);

    await run({ type: 'daily' });

    expect(sns.publish).toBeCalledTimes(0);
  });

  test('sends multiple email SNS notifications', async () => {
    db.getPingersByType.mockReturnValue([
      createPingerFixture({ id: 1 }),
      createPingerFixture({ id: 2 }),
    ]);
    db.getPropertyQueueForPingers.mockReturnValue([
      createPropertyQueueItemFixture({ id: 1, pinger_id: 1 }),
      createPropertyQueueItemFixture({ id: 2, pinger_id: 1 }),
      createPropertyQueueItemFixture({ id: 3, pinger_id: 2 }),
    ]);

    await run({ type: 'weekly' });

    expect(sns.publish).toBeCalledTimes(2);
  });

  test('locks and deletes used properties', async () => {
    db.getPingersByType.mockReturnValue([createPingerFixture()]);
    db.getPropertyQueueForPingers.mockReturnValue([
      createPropertyQueueItemFixture({ id: 1 }),
      createPropertyQueueItemFixture({ id: 2 }),
      createPropertyQueueItemFixture({ id: 3 }),
    ]);

    await run({ type: 'monthly' });

    expect(db.lockPropertyQueueItems).toBeCalledWith([1, 2, 3]);
    expect(db.deletePropertyQueueItems).toBeCalledWith([1, 2, 3]);
  });
});
