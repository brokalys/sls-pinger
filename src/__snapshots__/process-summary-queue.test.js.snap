// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process-summary-queue inserts an entry into the stats table 1`] = `
Array [
  1,
  Array [
    "122.32",
  ],
]
`;

exports[`process-summary-queue sends an email SNS notification 1`] = `
Array [
  Object {
    "Message": "email",
    "MessageAttributes": Object {
      "pinger_id": Object {
        "DataType": "Number",
        "StringValue": "1",
      },
      "subject": Object {
        "DataType": "String",
        "StringValue": "Jauni PINGER sludinƒÅjumi",
      },
      "template_id": Object {
        "DataType": "String",
        "StringValue": "summary",
      },
      "template_variables": Object {
        "DataType": "String",
        "StringValue": "{\\"is_premium\\":true,\\"limit_reached\\":false,\\"unsubscribe_url\\":\\"https://unsubscribe.brokalys.com/?key=undefined&id=c8bf6e7b-7eba-11eb-b2a8-663c33f40218\\",\\"properties\\":[[\\"https://brokalys.com/\\",100000,2,45,122.32,\\"https://brokalys.com/#/56.9965,24.245176,18/locate-building\\"],[\\"https://brokalys.com/\\",100000,2,45,122.32,\\"https://brokalys.com/#/56.9965,24.245176,18/locate-building\\"],[\\"https://brokalys.com/\\",100000,2,45,122.32,\\"https://brokalys.com/#/56.9965,24.245176,18/locate-building\\"]]}",
      },
      "to": Object {
        "DataType": "String",
        "StringValue": undefined,
      },
    },
    "MessageStructure": "string",
    "TargetArn": "arn:aws:sns:eu-west-1:111111111111:email-dev",
  },
]
`;
